lib/assert.js

```diff
@@ -931,6 +934,79 @@ assert.doesNotMatch = function doesNotMatch(string, regexp, message) {
   internalMatch(string, regexp, message, doesNotMatch);
 };
 
+const noop = () => {};
+
+const mustCallChecks = new Set();
+
+let attachExitHandler = true;
+
+function runCallChecks(exitCode) {
+  for (const context of mustCallChecks) {
+    if (context.actual !== context.exact) {
+      context.error = `Mismatched ${context.name} function calls. ` +
+        `Expected exactly ${context.exact}, actual ${context.actual}.`;
+      const console = require('console');
+      console.error(context.error);
+      console.error(context.stackTrace);
+      if (exitCode === 0)
+        process.exitCode = 1;
+    }
+  }
+}
+
+assert.mustCall = function mustCall(fn, exact = 1) {
+  if (process._exiting)
+    throw new ERR_MUSTCALL_UNAVAILABLE_DURING_EXIT();
+  if (typeof fn === 'number') {
+    exact = fn;
+    fn = noop;
+  } else if (fn === undefined) {
+    fn = noop;
+  }
+
+  validateUint32(exact, 'exact', true);
+
+  const context = {
+    exact,
+    actual: 0,
+    // eslint-disable-next-line no-restricted-syntax
+    stackTrace: new Error(),
+    name: fn.name || 'mustCall'
+  };
+
+  // Add the exit listener only once to avoid listener leak warnings
+  if (attachExitHandler === true) {
+    attachExitHandler = false;
+    process.on('exit', runCallChecks);
+  }
+
+  mustCallChecks.add(context);
+  return function() {
+    // Increment call count
+    context.actual++;
+    if (context.actual === context.exact) {
+      // Once function has reached its call count remove it from
+      // MustCallChecks set to prevent memory leaks
+      mustCallChecks.delete(context);
+    }
+    if (context.actual > context.exact) {
+      // If call count exceeds exact throw an error
+      const console = require('console');
+      console.error('Mismatched %s function calls.' +
+                    'Expected exactly %s, actual %d.',
+                    context.name,
+                    context.exact,
+                    context.actual);
+      console.error(context.stackTrace);
+      throw new ERR_MUSTCALL_CALL_COUNT_EXCEEDED(
+        context.name,
+        context.exact,
+        context.actual);
+    }
```

Comment:
I’m not sure if we really should throw an exception here…

- It looks like the exception comes from the function itself, which makes it possible that this exception will simply be caught by some other part of the code.
- As you mentioned, if this is uncaught, it interacts with the exit handler here in an odd way.

I would suggest just putting it back in the `mustCallChecks` list and letting the exit handler print a message about this.

```suggestion
    if (context.actual === context.exact + 1) {
      mustCallChecks.add(context);
```

Either way, what we definitely should not do imo is to both log something to stderr *and* throw an exception.

Reply:
I cleaned that block up there and removed the console.error lines